File: src/game.go
package src

// Game はオセロゲームの状態を管理する型
type Game struct {
	board         *Board
	currentPlayer CellState
}

// NewGame は新しいゲームを作成する
func NewGame() *Game {
	board := NewBoard()
	board.SetupInitialPosition()
	return &Game{
		board:         board,
		currentPlayer: Black, // 黒が先手
	}
}

// CurrentPlayer は現在の手番のプレイヤーを返す
func (g *Game) CurrentPlayer() CellState {
	return g.currentPlayer
}

// SwitchPlayer は手番を交代する
func (g *Game) SwitchPlayer() {
	if g.currentPlayer == Black {
		g.currentPlayer = White
	} else {
		g.currentPlayer = Black
	}
}

// CountStones は黒石と白石の数をカウントする
func (g *Game) CountStones() (blackCount, whiteCount int) {
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			pos := Position{Col: col, Row: row}
			cell := g.board.Get(pos)
			if cell == Black {
				blackCount++
			} else if cell == White {
				whiteCount++
			}
		}
	}
	return blackCount, whiteCount
}

// IsGameOver はゲームが終了したかどうかを判定する
func (g *Game) IsGameOver() bool {
	// 両者とも手を打てない場合、ゲーム終了
	return !g.CanPlayerMove(Black) && !g.CanPlayerMove(White)
}

// GetWinner は勝者を返す（引き分けの場合は Empty）
func (g *Game) GetWinner() CellState {
	blackCount, whiteCount := g.CountStones()
	if blackCount > whiteCount {
		return Black
	} else if whiteCount > blackCount {
		return White
	}
	return Empty // 引き分け
}

// CanPlayerMove は指定プレイヤーが手を打てるかどうかを判定する
func (g *Game) CanPlayerMove(player CellState) bool {
	moves := g.board.GetValidMoves(player)
	return len(moves) > 0
}

// PlaceStone は現在のプレイヤーが指定位置に石を置く
func (g *Game) PlaceStone(pos Position) {
	g.board.PlaceStone(pos, g.currentPlayer)
}

// GetBoard は盤面を返す（テスト用）
func (g *Game) GetBoard() *Board {
	return g.board
}

File: frontend/src/logic/game.ts
import type { Board, Color, GameState } from '../types/game';
import { createInitialBoard, getValidMoves, placeStone, countStones, isGameOver, getWinner } from './board';

export function createInitialGameState(): GameState {
  const board = createInitialBoard();
  return {
    board,
    currentPlayer: 'black',
    isGameOver: false,
    playId: null,
    blackCount: 2,
    whiteCount: 2,
    winner: null,
    validMoves: getValidMoves(board, 'black'),
    lastMove: null,
  };
}

function nextPlayer(current: Color): Color {
  return current === 'black' ? 'white' : 'black';
}

export function applyOpponentMove(state: GameState, row: number, col: number, color: Color): GameState {
  const newBoard = placeStone(state.board, row, col, color);
  if (newBoard === state.board) {
    return state;
  }
  return resolveNextTurn(newBoard, nextPlayer(color), { row, col });
}

export function handleMove(state: GameState, row: number, col: number): GameState {
  const newBoard = placeStone(state.board, row, col, state.currentPlayer);
  if (newBoard === state.board) {
    return state;
  }

  return resolveNextTurn(newBoard, nextPlayer(state.currentPlayer), { row, col });
}

function resolveNextTurn(board: Board, next: Color, lastMove: { row: number; col: number }): GameState {
  const { black, white } = countStones(board);

  if (isGameOver(board)) {
    return {
      board,
      currentPlayer: next,
      isGameOver: true,
      playId: null,
      blackCount: black,
      whiteCount: white,
      winner: getWinner(board),
      validMoves: [],
      lastMove,
    };
  }

  const nextMoves = getValidMoves(board, next);
  if (nextMoves.length > 0) {
    return {
      board,
      currentPlayer: next,
      isGameOver: false,
      playId: null,
      blackCount: black,
      whiteCount: white,
      winner: null,
      validMoves: nextMoves,
      lastMove,
    };
  }

  // Pass: the next player has no valid moves, so skip back
  const passedPlayer = nextPlayer(next);
  const passedMoves = getValidMoves(board, passedPlayer);
  return {
    board,
    currentPlayer: passedPlayer,
    isGameOver: false,
    playId: null,
    blackCount: black,
    whiteCount: white,
    winner: null,
    validMoves: passedMoves,
    lastMove,
  };
}
